### PR TITLE
Fixing issue in Voxel clearing (issue #662)

### PR DESCRIPTION
**Pull request by facontidavide**
Hi,

this commit https://github.com/ros-planning/navigation/commit/435a5a762036958e099bf5873b4bbccf82e39660 should solve the problem reported.

The idea is simple: use ALL the valid rays to clear.

The second commit is purely a pedantic "code simplification" that you can just ignore if you want.

Clearing points were discarded because either their height Z are outside the marking limits or because the Z coordinate of the point exceeded the size_z_ of the grid.

You can compare the results before:

https://vimeo.com/255893590

and after the changes

https://vimeo.com/256061162

You will noticed that there are still few voxels which are not cleared. The reason is that some rays of the velodyne are not hitting any object (they are pointing at the sky), therefore they are discarded by the velodyne_pointcloud driver.
In other words, PointCloud and VoxelLayer do not provide the equivalent of "infinite is valid", for this reason there are still marked voxel that can not be cleared.

### CODE REVIEW

**Comment by mikeferguson** [comment]
Davide, I've [commented in the issue](https://github.com/ros-planning/navigation/issues/662#issuecomment-376565069) that I think this can be accomplished through the configuration parameters available today. If that is the case, that is preferable, since it has less chance of possibly breaking other user's configurations. I'm going to add a few notes here and close this pull request to let dialogue continue in the issue. We can reopen if it turns out we really do need to modify things. Thanks!

File: costmap_2d/include/costmap_2d/observation.h [hunk 1]
```diff
@@ -97,6 +98,7 @@ class Observation
   geometry_msgs::Point origin_;
   pcl::PointCloud<pcl::PointXYZ>* cloud_;
   double obstacle_range_, raytrace_range_;
+  std::vector<bool> within_range_;
```
Comment (mikeferguson) [inline costmap_2d/include/costmap_2d/observation.h:101]: rather than creating a possibly very large bool array, I would suggest that we should instead pass the min_obstacle_height_ and max_obstacle_height_ into the Observation from the ObservationBuffer (as we already do for raytrace_range_ and obstacle_range_)

File: costmap_2d/plugins/voxel_layer.cpp [hunk 2]
```diff
@@ -146,32 +146,34 @@ void VoxelLayer::updateBounds(double robot_x, double robot_y, double robot_yaw,
...
+    const std::vector<bool>& in_range = obs.within_range_;
 
     double sq_obstacle_range = obs.obstacle_range_ * obs.obstacle_range_;
 
     for (unsigned int i = 0; i < cloud.points.size(); ++i)
     {
+      const pcl::PointXYZ& point = cloud.points[i];
       // if the obstacle is too high or too far away from the robot we won't add it
-      if (cloud.points[i].z > max_obstacle_height_)
+      if (in_range[i] == false || point.z > max_obstacle_height_ )
```
Comment (mikeferguson) [inline costmap_2d/plugins/voxel_layer.cpp:157]: this could then be updated to compare against the obs.max_obstacle_height_

File: costmap_2d/plugins/voxel_layer.cpp [hunk 3]
```diff
@@ -313,10 +315,11 @@ void VoxelLayer::raytraceFreespace(const Observation& clearing_observation, doub
     double t = 1.0;
 
     // we can only raytrace to a maximum z height
-    if (wpz > max_obstacle_height_)
+    const double max_z = std::min(max_obstacle_height_, size_z_ * z_resolution_ );
```
Comment (mikeferguson) [inline costmap_2d/plugins/voxel_layer.cpp:318]: I think we need to look closer at this specific item, as mentioned in the issue.

File: costmap_2d/include/costmap_2d/observation.h (hunk 1 above)
Comment (facontidavide) [reply costmap_2d/include/costmap_2d/observation.h:101]: I agree, we can do that

File: costmap_2d/plugins/voxel_layer.cpp (hunk 3 above)
Comment (facontidavide) [reply costmap_2d/plugins/voxel_layer.cpp:318]: From my point of view, either we apply this change or we simply remove the parameter __z_voxels__, since this value is supposed to be (max_obstacle_height / z_resolution)

**Comment by facontidavide** [comment]
Let's agree that we disagree :)

I guess that the best way to "convince you" that these changes are necessary might be to build a unit test that show clearly the behavior of the VoxelGrid before and after the PR.

Does it make sense?
